tool/tsh/db.go

```diff
@@ -348,38 +378,109 @@ func pickActiveDatabase(cf *CLIConf) (*tlsca.RouteToDatabase, error) {
+
+func WithLocalProxyRoute(host string, port int) ConnectCommandFunc {
+	return func(opts *connectionCommandOpts) {
+		opts.localProxyPort = port
+		opts.localProxyHost = host
+	}
+}
+
+func getConnectCommand(cf *CLIConf, tc *client.TeleportClient, profile *client.ProfileStatus, db *tlsca.RouteToDatabase, opts ...ConnectCommandFunc) (*exec.Cmd, error) {
+	var options connectionCommandOpts
+	for _, opt := range opts {
+		opt(&options)
+	}
+
 	switch db.Protocol {
 	case defaults.ProtocolPostgres:
-		return getPostgresCommand(db, profile.Cluster, cf.DatabaseUser, cf.DatabaseName), nil
+		return getPostgresCommand(db, profile.Cluster, cf.DatabaseUser, cf.DatabaseName, options), nil
 	case defaults.ProtocolMySQL:
-		return getMySQLCommand(db, profile.Cluster, cf.DatabaseUser, cf.DatabaseName), nil
+		return getMySQLCommand(db, profile.Cluster, cf.DatabaseUser, cf.DatabaseName, options), nil
 	case defaults.ProtocolMongoDB:
 		host, port := tc.WebProxyHostPort()
+		if options.localProxyPort != 0 && options.localProxyHost != "" {
+			host = options.localProxyHost
+			port = options.localProxyPort
+
+		}
 		return getMongoCommand(host, port, profile.DatabaseCertPath(db.ServiceName), cf.DatabaseName), nil
 	}
 	return nil, trace.BadParameter("unsupported database protocol: %v", db)
 }
 
-func getPostgresCommand(db *tlsca.RouteToDatabase, cluster, user, name string) *exec.Cmd {
+func mkLocalProxy(ctx context.Context, remoteProxyAddr string, protocol string, listener net.Listener) (*alpnproxy.LocalProxy, error) {
+	alpnProtocol, err := toALPNProtocol(protocol)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	address, err := utils.ParseAddr(remoteProxyAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+
+	}
+	lp, err := alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
+		RemoteProxyAddr: remoteProxyAddr,
+		Protocol:        alpnProtocol,
+		Listener:        listener,
+		ParentContext:   ctx,
+		SNI:             address.Host(),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return lp, nil
+}
+
+func toALPNProtocol(dbProtocol string) (alpnproxy.Protocol, error) {
+	switch dbProtocol {
+	case defaults.ProtocolMySQL:
+		return alpnproxy.ProtocolMySQL, nil
+	case defaults.ProtocolPostgres:
+		return alpnproxy.ProtocolPostgres, nil
+	case defaults.ProtocolMongoDB:
+		return alpnproxy.ProtocolMongoDB, nil
+	default:
+		return "", trace.NotImplemented("%q protocol not supported", dbProtocol)
+	}
+}
+
+func getPostgresCommand(db *tlsca.RouteToDatabase, cluster, user, name string, options connectionCommandOpts) *exec.Cmd {
 	connString := []string{fmt.Sprintf("service=%v-%v", cluster, db.ServiceName)}
 	if user != "" {
 		connString = append(connString, fmt.Sprintf("user=%v", user))
 	}
 	if name != "" {
 		connString = append(connString, fmt.Sprintf("dbname=%v", name))
 	}
+	if options.localProxyPort != 0 {
+		connString = append(connString, fmt.Sprintf("--port=%v", options.localProxyPort))
```

Comment:
Instead of flags, let's use connection string parameters - `host=%v` and `port=%v`. I've had some issues in the past trying to mix flags and connection string.